src/storage/heap_file.c

```diff
@@ -23288,17 +23311,20 @@ heap_hfid_cache_get (THREAD_ENTRY * thread_p, const OID * class_oid, HFID * hfid
       /* this is either a newly inserted entry or one with incomplete information that is currently being filled by
        * another transaction. We need to retrieve the HFID from the class record. We do not care that we are
        * overwriting the information, since it must be always the same (the HFID never changes for the same class OID). */
-      error_code = heap_get_hfid_from_class_record (thread_p, class_oid, &hfid_local);
+      error_code = heap_get_hfid_from_class_record (thread_p, class_oid, &hfid_local, true, &classname_local);
       if (error_code != NO_ERROR)
 	{
 	  ASSERT_ERROR ();
 	  lf_tran_end_with_mb (t_entry);
 	  return error_code;
 	}
       entry->hfid = hfid_local;
+      entry->classname = classname_local;
```

Comment:
This operation is not thread safe. There can be several threads racing here to set information for the first time. If setting hfid/file_type multiple times cannot cause harm, setting classname multiple times can cause memory leaks.

Make entry->classname an atomic pointer. Setting classname needs compare_exchange operation from NULL to new value.